engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceManager.java

```diff
@@ -773,67 +795,73 @@ public Map<String, Object> getTableAttributes(
             this.tle = tle;
-            this.rowSetSlice = rowSetSlice;
+            this.rowSet = rowSet;
         }
 
         @Override
         public void close() {
-            rowSetSlice.close();
+            rowSet.close();
         }
     }
 
     /**
-     * Get the included regions and corresponding row set slice that overlap with the input row set. The user is
-     * responsible for closing the returned row sets.
+     * Return the first {@code numRegions} regions that overlap with the input row set, and their corresponding
+     * overlapping row sets. The user is responsible for closing the returned objects.
+     */
+    private ArrayList<RegionInfoHolder> getOverlappingRegions(final RowSet inputRowSet, final int numRegions) {
+        final List<IncludedTableLocationEntry> tableLocationEntries = includedLocationEntries();
+        final ArrayList<RegionInfoHolder> includedRegions = new ArrayList<>(numRegions);
+        try (final RowSet.SearchIterator it = inputRowSet.searchIterator()) {
+            while (includedRegions.size() < numRegions && it.hasNext()) {
+                final long startKey = it.nextLong();
+                final int regionIndex = RegionedColumnSource.getRegionIndex(startKey);
+                if (regionIndex >= tableLocationEntries.size()) {
+                    throw new IllegalStateException("Region index " + regionIndex + " exceeds the number of included " +
+                            "locations: " + tableLocationEntries.size() + " for input row set: " + inputRowSet);
+                }
+                final IncludedTableLocationEntry entry = tableLocationEntries.get(regionIndex);
+                try (final WritableRowSet overlappingRowSet = entry.getOverlappingRowSet(inputRowSet)) {
+                    if (overlappingRowSet.isEmpty()) {
+                        throw new IllegalStateException(
+                                "Overlapping row set is empty for region index " + regionIndex + " and input row set: "
+                                        + inputRowSet);
+                    }
+                    includedRegions.add(new RegionInfoHolder(entry, overlappingRowSet.copy()));
+                    final long endKey = overlappingRowSet.lastRowKey();
+                    it.advance(endKey); // Move to the next region, skipping the current one
+                }
+            }
+        }
+        return includedRegions;
+    }
+
+    /**
+     * Get all included regions and corresponding row set that overlap with the input row set. The user is responsible
+     * for closing the returned row sets.
      *
      * @param inputRowSet The input row set
      * @return A list of included region information
      */
-    private ArrayList<OverlappingRegionInfo> getOverlappingRegions(final RowSet inputRowSet) {
-        final ArrayList<OverlappingRegionInfo> includedRegions = new ArrayList<>();
+    private ArrayList<RegionInfoHolder> getOverlappingRegions(final RowSet inputRowSet) {
```

Comment:
Should we use the search iterator approach here too? (We can probably just delegate to it w/ `numRegions=Integer.MAX_VALUE`?)